pyvisa_py/prologix.py

```diff
@@ -0,0 +1,397 @@
+        )
+        # self.interface.write_termination = "\n"
+        # self.write_termination = "\n"
+
+        for name in (
+            "ASRL_END_IN",
+            "ASRL_END_OUT",
+            "SEND_END_EN",
+            "TERMCHAR",
+            "TERMCHAR_EN",
+            "SUPPRESS_END_EN",
+        ):
+            attribute = getattr(constants, "VI_ATTR_" + name)
+            self.attrs[attribute] = attributes.AttributesByID[attribute].default
+
+    def write(self, data: bytes) -> Tuple[int, StatusCode]:
+        """Writes data to device or interface synchronously.
+
+        Corresponds to viWrite function of the VISA library.
+
+        Parameters
+        ----------
+        data : bytes
+            Data to be written.
+
+        Returns
+        -------
+        int
+            Number of bytes actually transferred
+        StatusCode
+            Return value of the library call.
+
+        """
+        if self.interface is None:
+            raise errors.InvalidSession()
+
+        if self.interface.inWaiting() > 0:
+            self.interface.flushInput()
+        self.plus_plus_read = True
+        return super().write(data)
+
+
+class PrologixInstrSession(Session):
+    """
+    This class is instantiated for GPIB<n>::INSTR resources, but only when
+    the corresponding PRLGX-xxx<n>::INTFC resource has been instantiated.
+    """
+
+    # we don't decorate this class with Session.register() because we don't
+    # want it to be registered in the _session_classes array, but we still
+    # need to define session_type to make the set_attribute machinery work.
+    session_type = (constants.InterfaceType.gpib, "INSTR")
+
+    # Override parsed to take into account the fact that this
+    # class is only used for a specific kind of resource
+    parsed: rname.GPIBInstr
+
+    @staticmethod
+    def list_resources() -> List[str]:
+        # TODO: is there a way to get this?
+        return []
+
+    def after_parsing(self) -> None:
+        self.interface = BOARDS[self.parsed.board]
+        self.gpib_addr = self.parsed.primary_address
+        if self.parsed.secondary_address:
+            # Secondary address of the device to connect to
+            # Reference for the GPIB secondary address
+            # https://www.mathworks.com/help/instrument/secondaryaddress.html
+            # NOTE: a secondary address of 0 is not the same as no secondary address.
+            self.gpib_addr += " " + self.parsed.secondary_address
+
+    def close(self) -> StatusCode:
+        if self.interface is None or self.interface.interface is None:
+            raise errors.InvalidSession()
+
+        self.interface = None
+        return StatusCode.success
+
+    def read(self, count: int) -> Tuple[bytes, StatusCode]:
+        if self.interface is None or self.interface.interface is None:
+            raise errors.InvalidSession()
+
+        self.interface.gpib_addr = self.gpib_addr
+
+        return self.interface.read(count)
```

Comment:
Looking at this code I think we need a lock on the interface.

Reply:
Good catch, I've added a lock